SwissTransferFeatures/UploadProgressView/TransferSessionManager.swift

```diff
@@ -104,58 +92,116 @@ class TransferSessionManager: ObservableObject {
 
-            try await uploadFile(atPath: localFile.localPath, toRemoteFile: remoteUploadFile, uploadUUID: uploadSession.uuid)
-        }
+        try await remoteUploadFiles.enumerated()
+            .map { (uploadWithRemoteContainer.files[$0.offset], $0.element) }
+            .asyncForEach { localFile, remoteUploadFile in
+                try await transferManagerWorker.uploadFile(
+                    atPath: localFile.localPath,
+                    remoteUploadFileUUID: remoteUploadFile.uuid,
+                    uploadUUID: uploadSession.uuid
+                )
+            }
 
         Logger.general.info("Found container: \(container.uuid)")
 
         let transferUUID = try await uploadManager.finishUploadSession(uuid: uploadSession.uuid)
 
         return transferUUID
     }
+}
+
+struct TransferManagerWorker {
+    private static let maxParallelUploads = 4
+    private let uploadURLSession = URLSession.shared
+
+    private let rangeProviderConfig = RangeProvider.Config(
+        chunkMinSize: 50 * 1024 * 1024,
+        chunkMaxSizeClient: 50 * 1024 * 1024,
+        chunkMaxSizeServer: 50 * 1024 * 1024,
+        optimalChunkCount: 200,
+        maxTotalChunks: 10000,
+        minTotalChunks: 1
+    )
+
+    let overallProgress: Progress
 
-    private func uploadFile(atPath: String, toRemoteFile: any RemoteUploadFile, uploadUUID: String) async throws {
-        guard let fileURL = URL(string: atPath) else {
-            throw ErrorDomain.invalidURL(rawURL: atPath)
+    func uploadFile(atPath: String, remoteUploadFileUUID: String, uploadUUID: String) async throws {
+        guard let fileURL = URL(string: atPath),
+              let chunkReader = ChunkReader(fileURL: fileURL) else {
+            throw TransferSessionManager.ErrorDomain.invalidURL(rawURL: atPath)
         }
 
-        let rangeProvider = RangeProvider(fileURL: fileURL, config: TransferSessionManager.rangeProviderConfig)
+        let rangeProvider = RangeProvider(fileURL: fileURL, config: rangeProviderConfig)
 
         let ranges = try rangeProvider.allRanges
-        guard let chunkProvider = ChunkProvider(fileURL: fileURL, ranges: ranges) else {
-            throw ErrorDomain.invalidRangeCompute
-        }
 
-        let rangeCount = ranges.reduce(0) { $0 + $1.count }
-        let fileProgress = Progress(totalUnitCount: Int64(rangeCount))
-        overallProgress?.addChild(fileProgress, withPendingUnitCount: Int64(rangeCount))
-
-        var index: Int32 = 0
-        while let chunk = chunkProvider.next() {
-            guard let rawChunkURL = try injection.sharedApiUrlCreator.uploadChunkUrl(
-                uploadUUID: uploadUUID,
-                fileUUID: toRemoteFile.uuid,
-                chunkIndex: index,
-                isLastChunk: index == ranges.count - 1
-            ) else {
-                throw ErrorDomain.invalidUploadChunkURL
+        try await ranges
+            .dropLast()
```

Comment:
If you use a `popLast`  earlier instead, you will not have to rely on array index math to get the last chunk interval. You will just get it from here directly.